### ISSUE TITLE
Reducing Keystrokes

### ISSUE THREAD

**Issue by christophercrouzet** [lesspass/lesspass#386]
Hello,

I only recently started looking into password managers and decided to choose LessPass in the process, for which I am really happy with so far, and grateful for.

Using LessPass (and probably other password managers?) takes a bit more time to log into each account but I understand and accept it. That being said, I would like to share with you a couple of UX ideas that could possibly help minimizing this drawback.

First, I've read https://github.com/lesspass/lesspass/issues/258 and, for the `Site` field, I personally prefer using only the domain (e.g. _google_) rather than the FQDN (e.g. _accounts.google.com_), so I very much appreciate the suggestion box that has been implemented. Alas, in the end, I often find myself filling the `Site` field manually rather than using the suggestion box.

For context, I like to avoid as much as possible using the mouse to speed-up my "workflow", so here are the actions required for me to make use of the suggestion box:

- hit <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>L</kbd> to open the pop-up.
- hit <kbd>Shift</kbd> + <kbd>Tab</kbd> to put the focus on the `Site` field since it defaults to the `Login` field.
- hit <kbd>→</kbd> to put the cursor at the end of the `Site` field content.
- hit <kbd>←Backspace</kbd> to delete the last character and trigger the suggestion box.
- navigate using <kbd>↑</kbd> or <kbd>↓</kbd> until the appropriate suggestion is selected.

Since the whole content of the `Site` field is selected when hitting <kbd>Shift</kbd> + <kbd>Tab</kbd>, I find it more intuitive to direclty start typing the content (e.g. _google_) rather than shifting the position of my hands to trigger the suggestion box. What would really encourage me using the suggestion box though is if it showed up whenever I'd start typing. For example, if I'd type the letter 'a', then _accounts_, _accounts.google_, and _accounts.google.com_ would be suggested. As for the letter 'g', only _google_ and _google.com_ would be suggested.

Better would be to have, additionally to the suggestion box, an autofill for the `Site` field. But probably as a setting disabled by default, which could be enabled through the add-on preferences.

Even better would be to have yet another add-on preference where we could provide custom `Site` and `Login` defaults for each FQDN. I understand that this could potentially contribute to reducing security but I believe that it is fine to let the users decide how they want to balance security versus usability. For example, I am personally happy with having the `Master Password` as the only salt that only me knows, if it can help me to log-in faster.

Secondly, I find it tedious to, each time, press the `Generate` button, then the `Copy` button, and then hit <kbd>Esc</kbd>, to finally being able to paste the pasword in the browser, when these 3 actions could potentially be merged into one. I understand that I might overlook some usage scenarios but it would be nice to have a new setting in the add-on preferences that provides different behaviours upon pressing the `Generate` button. The values available to choose upon would be `Generate Only` (the current default), `Generate and Copy`, and `Generate, Copy, and Close`. I see this issue mentioned in https://github.com/lesspass/lesspass/issues/123, but maybe this proposed approach could help with the limitation of clipboardjs? Or maybe not, I'm not familiar with the problem :stuck_out_tongue:

## Additional Notes

- in Firefox, using the shortcut <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>L</kbd> won't work if the add-on is not visible in the toolbar. Maybe it is a restriction stemming from Firefox itself but it is a bit of a bummer since having the icon visible makes it obvious to anyone seeing my screen that I am using a password manager.
- I find the term `Site` to be slightly misleading. This word refers to a location or, in the context of Internet, to a web page, which could lead to thinking that LessPass is only made to be used in the context of a browser. Yet, I am happy to use it for desktop applications such as Skype or Steam, so maybe a more generic term like `Entry Name` would be more appropriate? After all, some users probably already use entry names such as _Micro$oft_ to log into their account on the hotmail.com site, just to obfuscate things a bit more.

Thank you for your work!

**Comment by panther2** [comment]
@christophercrouzet 

> ... I am happy to use it for desktop applications such as Skype or Steam, so maybe a more generic term like `Entry Name` would be more appropriate?

Well said, LessPass is much more than "only" a website-tool. I use it for **everything** that requires passwords.

**Comment by guillaumevincent** [comment]
Hello,
thank you for this issue :heart: 

You are addressing a lot of points.

 * autocompletion: yes we should create a configurable behavior to help user to have the better experience. I personally remove the `www.` all the time.
 * copy after generation is mandatory only because of Firefox (Firefox dev if you read me!) Firefox copy api refuse to copy text in an asynchronous function. This is absurd but we can't do anything about this one (Except to have a different behavior between Firefox and others browsers)
 * <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>L</kbd> on Firefox when not in the toolbar is a Firefox issue also. Maybe I can try to open or search a BZ for this one.
 * `Site` term is good for newcomers. Maybe we can try to do some ux research for this one. This is time consuming. I don't know if I will have some time for this one.

To adress this issue we probably can try to work on site autocompletion. The user experience can be better. Based on your comment here an idea:

You arrive on `https://www.example.org`

 * <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>L</kbd> to open LessPass
 * site field is `www.example.org` and the text in the site field is selected. You can choose:
    * type the letter <kbd>e</kbd> to remove the text and show the auto completion with `example` and `example.org`. Then hit <kbd>↑</kbd> or <kbd>↓</kbd> to select the completion.
    * hit <kbd>tab</kbd> to keep `www.example.org` and move to the next field

any idea @edouard-lopez on this one?

**Comment by christophercrouzet** [comment]
Thanks for taking the time to reply, @guillaumevincent!

> copy after generation is mandatory only because of Firefox (Firefox dev if you read me!) Firefox copy api refuse to copy text in an asynchronous function. This is absurd but we can't do anything about this one (Except to have a different behavior between Firefox and others browsers)

By no mean do I understand the specifics here and I trust you to know (much) better but I couldn't help looking at the code out of curiosity and found both the `generatePassword()` and the `copyPassword()` functions from the lesspass-pure module. Module which is sadly not available on GitHub anymore?

From there, it seems that you are using the copy-text-to-clipboard module to copy the generated password and, in turn, this module relies on calling `document.execCommand('copy')`.

When reading [Firefox's Clipboard API page](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API), they mention that “this API is designed to supersede accessing the clipboard using document.execCommand()”. They also refer to the [Clipboard Interface](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard) as being async. So maayyybee, the [`navigator.clipboard.writeText()`](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/writeText) function could work in the `then` context of your `LessPass.generatePassword()` function call?


> <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>L</kbd> on Firefox when not in the toolbar is a Firefox issue also. Maybe I can try to open or search a BZ for this one.

Sounds good... wait, what is a BZ? :stuck_out_tongue:


> `Site` term is good for newcomers. Maybe we can try to do some ux research for this one. This is time consuming. I don't know if I will have some time for this one.

Up to you! For context though, when I first opened LessPass and saw the term `Site`, I was concerned about my password being somehow tied to the URL of the site that I tried to log into, and about what would happen if they decided to change the login URL one day? I had to look into the `_calc_entropy()` function to realize that it was simply a salt and that I was free to put whatever worked best for me. Admittedly, it was probably mentioned somewhere but I overlooked all sort of documentation—after all, good UIs don't need documentation anyways, right? :stuck_out_tongue:


> You arrive on `https://www.example.org`
> 
> * <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>L</kbd> to open LessPass
> * site field is `www.example.org` and the text in the site field is selected. You can choose:
>   * type the letter e to remove the text and show the auto completion with `example` and `example.org`. Then hit <kbd>↑</kbd> or <kbd>↓</kbd> to select the completion.
>   * hit <kbd>tab</kbd> to keep `www.example.org` and move to the next field
  
I would very much love to have this! I am only concerned with the users who keep the default site and are used to directly start typing in the login field. So, to not disturb their experience, it might be worth adding an add-on preference whether the site or login field should be selected by default?

I'd still rather have an add-on preference with custom defaults for each FQDN, even though I concur that it's a bit ugly. But it was done before, for example for the rules/whitelists in the uBlock Origin add-on, and it's pretty useful since it'd allow users to only type the master password for the entries where defaults for the site & login fields were already provided. And hey, it could even be used as yet another obfuscation method for the more paranoids of us by purposedly entering wrong defaults :smile:


Cheers!
